src/daemon/distribution.c

```diff
@@ -0,0 +1,305 @@
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+distribution_t *distribution_new_custom(size_t num_bounds,
+                                        double *custom_max_boundaries) {
+
+  if ((num_bounds == 0) || (custom_max_boundaries == NULL)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  for (size_t i = 1; i < num_bounds; i++) {
+    if ((custom_max_boundaries[i] <= custom_max_boundaries[i - 1]) ||
+        (custom_max_boundaries[i] == INFINITY)) {
+      errno = EINVAL;
+      return NULL;
+    }
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets =
+      calloc(num_bounds + 1, sizeof(bucket_t)); //+1 for infinity bucket
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_bounds + 1; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] =
+          initialize_bucket(0, custom_max_boundaries[0]);
+    } else if (i < num_bounds) {
+      new_distribution->buckets[i] =
+          initialize_bucket(new_distribution->buckets[i - 1].max_boundary,
+                            custom_max_boundaries[i]);
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets =
+      num_bounds + 1; // plus one for infinity bucket
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+static size_t binary_search(distribution_t *dist, size_t left, size_t right,
+                            double gauge) {
+  if (left > right) {
+    return -1;
+  }
+
+  size_t mid = left + (right - left) / 2;
+  if (gauge >= dist->buckets[mid].min_boundary &&
+      gauge < dist->buckets[mid].max_boundary) {
+    return mid;
+  }
+
+  if (gauge < dist->buckets[mid].min_boundary) {
+    return binary_search(dist, left, mid - 1, gauge);
+  }
+
+  return binary_search(dist, mid + 1, right, gauge);
+}
+
+int distribution_update(distribution_t *dist, double gauge) {
+  if ((dist == NULL) || (gauge <= 0)) {
+    return EINVAL;
+  }
+  /*
+  for(size_t i = 0; i < dist->num_buckets; i++) {
+    if(gauge >= dist->buckets[i].min_boundary && gauge <
+  dist->buckets[i].max_boundary) { dist->buckets[i].bucket_counter++;
+    }
+  }
+  */
+  size_t left = 0;
+  size_t right = dist->num_buckets - 1;
+  size_t index = binary_search(dist, left, right, gauge);
+
+  dist->buckets[index].bucket_counter++;
+  dist->total_scalar_count++;
+  dist->raw_data_sum += gauge;
+  return 0;
+}
+
+double distribution_average(distribution_t *dist) {
+  if (dist == NULL) {
+    errno = EINVAL;
+    return NAN;
+  }
+
+  if (dist->total_scalar_count == 0) {
+    return NAN;
+  }
+
+  return dist->raw_data_sum / dist->total_scalar_count;
```

Comment:
I'd recommend to cast it to the double type. Without it it can work but some compilers can complain about that.

Reply:
Thank you Barbara, I agree with you.